### PR TITLE
Fix dockerd internal port changing on restart

### PR DESCRIPTION
**Pull request by medyagh**
closes https://github.com/kubernetes/minikube/issues/7017
### Gotcha 0 : driver.GetURL() is a big Lie !
GetURL in our driver is a very dirty hack, and has nothing to do with our Machine's URL it is just something to satify libmachine that wants to find out the insalled docker inside the node. 
so I had to make kic driver follow hyperkit driver and make GetURL to return a BIG fat lie !
and later replace it in setupKubeConfig function in start ...very hacky and dirty
but till we come up with a solution for machine driver. that is the solution.

related: https://github.com/docker/machine/blob/master/libmachine/provision/utils.go#L159_L175

### Gotcha 1 : Daemon reload

That was the ps aux params for docker in kic driver:
```
root@minikube:/home/docker# ps aux | grep dockerd
root     42776  2.3  1.2 2118696 98856 ?       Ssl  20:48   0:21 /usr/bin/dockerd -H fd:// --containerd=/run/containerd/containerd.sock
```

and this was for VM
```
root      2097  1.1  1.6 632732 65972 ?        Ssl  20:35   0:20 /usr/bin/dockerd -H tcp://0.0.0.0:2376 -H unix:///var/run/docker.sock --default-ulimit=nofile=1048576:1048576 --tlsverify --tlscacert /etc/docker/ca.pem --tlscert /etc/docker/server.pem --tlskey /etc/docker/server-key.pem --label provider=hyperkit --insecure-registry 10.96.0.0/12
```

after doing daemon-reload the KIC one also picked up the right settings.

### Gotcha 2: Cert pining was missing 127.0.01
when dockerd for kic has the right parameters (same as vm) it compains about cert not having the same IP
but the cert is generated by libmachine. so I had to turn of the tls verify for docker inside minikube
```
Error during connect: Get https://127.0.0.1:32799/v1.40/containers/json: x509: certificate is valid for 172.17.0.2, not 127.0.0.1
```

for that I had to Fork libmachine and add 127.0.0.1 in addition to "localhost" to the certs 
I created a PR in libmachine, till that PR gets merged, we are gonna use my fork because we need to release https://github.com/machine-drivers/machine/pull/28/files

@afbjorklund might be able to help

### Gotcha 3: Libmachuine expects Docker Port to be 2376
it is poiintless to have dockerport in a variable, because of libmachine it always has to be "2376" 
so I hard coded it for less confusing debugging for next person. to save some of their hair on their head.

### CODE REVIEW

File: pkg/provision/ubuntu.go [hunk 1]
```diff
@@ -155,6 +155,10 @@ WantedBy=multi-user.target
 		return nil, err
 	}
 
+	if err := p.Service("", serviceaction.DaemonReload); err != nil {
```
Comment (afbjorklund) [inline pkg/provision/ubuntu.go]: You can remove this, it is done automatically by serviceaction.Start

File: pkg/provision/buildroot.go [hunk 2]
```diff
@@ -100,7 +100,7 @@ Environment=DOCKER_RAMDISK=yes
...
 # container runtimes. If left unlimited, it may result in OOM issues with MySQL.
 ExecStart=
-ExecStart=/usr/bin/dockerd -H tcp://0.0.0.0:{{.DockerPort}} -H unix:///var/run/docker.sock --default-ulimit=nofile=1048576:1048576 --tlsverify --tlscacert {{.AuthOptions.CaCertRemotePath}} --tlscert {{.AuthOptions.ServerCertRemotePath}} --tlskey {{.AuthOptions.ServerKeyRemotePath}} {{ range .EngineOptions.Labels }}--label {{.}} {{ end }}{{ range .EngineOptions.InsecureRegistry }}--insecure-registry {{.}} {{ end }}{{ range .EngineOptions.RegistryMirror }}--registry-mirror {{.}} {{ end }}{{ range .EngineOptions.ArbitraryFlags }}--{{.}} {{ end }}
+ExecStart=/usr/bin/dockerd -H tcp://0.0.0.0:2376 -H unix:///var/run/docker.sock --default-ulimit=nofile=1048576:1048576 --tlsverify --tlscacert {{.AuthOptions.CaCertRemotePath}} --tlscert {{.AuthOptions.ServerCertRemotePath}} --tlskey {{.AuthOptions.ServerKeyRemotePath}} {{ range .EngineOptions.Labels }}--label {{.}} {{ end }}{{ range .EngineOptions.InsecureRegistry }}--insecure-registry {{.}} {{ end }}{{ range .EngineOptions.RegistryMirror }}--registry-mirror {{.}} {{ end }}{{ range .EngineOptions.ArbitraryFlags }}--{{.}} {{ end }}
```
Comment (afbjorklund) [inline pkg/provision/buildroot.go]: most likely, you don't have to modify this file

File: pkg/drivers/kic/kic.go [hunk 3]
```diff
@@ -202,11 +202,13 @@ func (d *Driver) GetSSHKeyPath() string {
...
 func (d *Driver) GetURL() (string, error) {
-	p, err := oci.HostPortBinding(d.NodeConfig.OCIBinary, d.MachineName, d.NodeConfig.APIServerPort)
-	url := fmt.Sprintf("https://%s", net.JoinHostPort("127.0.0.1", fmt.Sprint(p)))
+	ip, err := d.GetIP()
 	if err != nil {
-		return url, errors.Wrap(err, "api host port binding")
+		return "", err
 	}
+	// because of libmachine problem with detecting docker port https://github.com/kubernetes/minikube/issues/7017
+	// we have to do this hack https://github.com/docker/machine/blob/master/libmachine/provision/utils.go#L159_L175
```
Comment (tstromberg) [inline pkg/drivers/kic/kic.go]: It's unclear what the hack actually is - it looks fairly normal to me. A more useful comment would be:

`// oci.HostPortBinding can be wrong, so hardcode the port because of ...`

I like the links to the issue, but don't understand the link to the code. Also, the code link will go stale at some point, since it doesn't pin to a revision.

File: pkg/drivers/kic/kic.go [hunk 4]
```diff
@@ -202,11 +202,13 @@ func (d *Driver) GetSSHKeyPath() string {
 
 // GetURL returns ip of the container running kic control-panel
```
Comment (tstromberg) [inline pkg/drivers/kic/kic.go]: Update the comment here, as it isn't returning an IP. Maybe specifically say it returns a URL to the Docker endpoint to Docker inside of the container? It is inside, isn't it?

File: pkg/minikube/node/config.go [hunk 5]
```diff
@@ -127,12 +129,24 @@ func setupKubeAdm(mAPI libmachine.API, cfg config.ClusterConfig, node config.Nod
...
 func setupKubeconfig(h *host.Host, c *config.ClusterConfig, n *config.Node, clusterName string) (*kubeconfig.Settings, error) {
-	addr, err := h.Driver.GetURL()
-	if err != nil {
-		exit.WithError("Failed to get driver URL", err)
-	}
-	if !driver.IsKIC(h.DriverName) {
+	addr := ""
+	var err error
+	if driver.IsKIC(h.DriverName) {
+		p, err := oci.HostPortBinding(h.DriverName, h.Name, n.Port)
```
Comment (tstromberg) [inline pkg/minikube/node/config.go]: For another PR perhaps, but `HostPortBinding` may be a poor name for this function, as it sounds like it may be binding to a port rather than fetching information. Perhaps:

`oci.ForwardedPort` or `oci.BoundPort`, `oci.HostPort`?

File: pkg/minikube/node/config.go [hunk 6]
```diff
@@ -127,12 +129,24 @@ func setupKubeAdm(mAPI libmachine.API, cfg config.ClusterConfig, node config.Nod
...
+		addr, err = h.Driver.GetURL()
+		if err != nil {
+			exit.WithError("Failed to get driver URL", err)
+		}
+
 		addr = strings.Replace(addr, "tcp://", "https://", -1)
+		// because of libmachine problem with detecting docker port https://github.com/kubernetes/minikube/issues/7017
+		// we have to do this hack https://github.com/docker/machine/blob/master/libmachine/provision/utils.go#L159_L175
+		// contact @afbjorklund for more details
 		addr = strings.Replace(addr, ":2376", ":"+strconv.Itoa(n.Port), -1)
```
Comment (tstromberg) [inline pkg/minikube/node/config.go]: This code is getting hard to follow, and I especially think it would be cleaner to call GetIP instead of search and replacing hardcoded numbers. What do you think about extracting this logic into a function like:

```
func apiServerURL() {
  if KIC() {
     p = oci.HostPortBinding(h.DriverName, h.Name, n.Port)
     return fmt.Sprintf("https://" + net.JoinHostPort(oci.DefaultBindIPV4, strconv.Itoa(p))) 
  }
  ip := h.Driver.GetIP()
  return fmt.Sprintf("https://" + net.JoinHostPort(ip, strconv.Itoa(p)))
}
```

File: pkg/provision/buildroot.go [hunk 7]
```diff
@@ -151,6 +151,10 @@ WantedBy=multi-user.target
 		return nil, err
 	}
 
+	if err := p.Service("", serviceaction.DaemonReload); err != nil {
```
Comment (tstromberg) [inline pkg/provision/buildroot.go]: Now that the docker port is hardcoded, I think this can go away. If not, please add a comment describing why we need to reload systemd files.

File: pkg/minikube/node/config.go (hunk 5 above)
Comment (medyagh) [reply pkg/minikube/node/config.go]: created an issue for it

File: pkg/provision/buildroot.go (hunk 7 above)
Comment (medyagh) [reply pkg/provision/buildroot.go]: the reason is in kic we install it and it already has a service file

Comment (priyawadhwa) [reply pkg/provision/buildroot.go]: This step can take a couple seconds to run on my laptop -- if it's not necessary it would be better to remove it from a performance standpoint

File: pkg/minikube/node/config.go (hunk 6 above)
Comment (medyagh) [reply pkg/minikube/node/config.go]: good suggestion. done

File: pkg/minikube/node/config.go [hunk 8]
```diff
@@ -126,18 +129,13 @@ func setupKubeAdm(mAPI libmachine.API, cfg config.ClusterConfig, node config.Nod
...
 	}
-	if !driver.IsKIC(h.DriverName) {
-		addr = strings.Replace(addr, "tcp://", "https://", -1)
-		addr = strings.Replace(addr, ":2376", ":"+strconv.Itoa(n.Port), -1)
-	}
-
-	if c.KubernetesConfig.APIServerName != constants.APIServerName {
-		addr = strings.Replace(addr, n.IP, c.KubernetesConfig.APIServerName, -1)
+	if cc.KubernetesConfig.APIServerName != constants.APIServerName {
+		addr = strings.Replace(addr, n.IP, cc.KubernetesConfig.APIServerName, -1)
```
Comment (tstromberg) [inline pkg/minikube/node/config.go]: Now that the worst was refactored out, this search and replace is clearly broken, especially with KIC:

- The APIServerName is not guaranteed to be a hostname at all. It can be any arbitrary string.
- We're searching and replacing for an IP which may not exist in the string

I see this feature was added here: https://github.com/kubernetes/minikube/issues/3878

I don't totally love this feature, but my suggestion would be to move this logic to apiServerURL so that it can be done without search and replace. It probably isn't necessary to support with KIC, but the search and replace is currently broken in this use case, as n.IP == 172.17.0.x, and this URL uses 127.0.0.1.

File: pkg/provision/buildroot.go [hunk 9]
```diff
@@ -151,6 +151,11 @@ WantedBy=multi-user.target
 		return nil, err
 	}
 
+	// because currently we install docker differently in kic, it is harmless to do this step for VM too so the code by the same
```
Comment (tstromberg) [inline pkg/provision/buildroot.go]: I wouldn't go so far as claiming harmless - it does add startup latency and room for failure in a place where it isn't necessary.  I'm not convinced we should call this in the VM, but could easily be swayed by a good comment.

// We have possibly changed the Docker file, inform systemctl of this fact.

File: pkg/provision/ubuntu.go [hunk 10]
```diff
@@ -155,6 +155,11 @@ WantedBy=multi-user.target
 		return nil, err
 	}
 
+	// because in kic base image we pre-install docker it already has a service file, it is harmless to double check with this step
```
Comment (tstromberg) [inline pkg/provision/ubuntu.go]: I don't want harmless commands to be executed. Docker is a hermetic environment: either we need to run this command or not. Be clear that this has to be run to pick up the reconfigured docker file.

File: pkg/minikube/node/config.go (hunk 8 above)
Comment (medyagh) [reply pkg/minikube/node/config.go]: done

File: pkg/minikube/node/config.go [hunk 11]
```diff
@@ -156,6 +152,32 @@ func setupKubeconfig(h *host.Host, c *config.ClusterConfig, n *config.Node, clus
...
+	if driver.IsKIC(h.DriverName) {
+		p, err := oci.HostPortBinding(h.DriverName, h.Name, n.Port)
+		if err != nil {
+			return "", errors.Wrap(err, "host port binding")
+		}
+		// for kic drivers we use 127.0.0.1 instead of node IP,
+		// because of Docker on MacOs limitations for reaching to container's IP.
+		addr := fmt.Sprintf("https://" + net.JoinHostPort(oci.DefaultBindIPV4, strconv.Itoa(p)))
+		if cc.KubernetesConfig.APIServerName != constants.APIServerName {
+			addr = strings.Replace(addr, oci.DefaultBindIPV4, cc.KubernetesConfig.APIServerName, -1)
```
Comment (tstromberg) [inline pkg/minikube/node/config.go]: Please do not use string replacement. It makes the code more difficult to follow and adds room for bugs, as we've already seen. What do you think of something like:

```
hostname := h.Driver.GetIP()
port := n.Port

if driver.IsKIC() {
  hostname := oci.DefaultBindIPV4
  port := ...
}

// Warning: this may be an arbitrary string
if  cc.KubernetesConfig.APIServerName != constants.APIServerName {
  hostname =  cc.KubernetesConfig.APIServerName
}

return fmt.Sprintf("https://" + net.JoinHostPort(hostname, strconv.Itoa(port)))
```

File: pkg/minikube/node/config.go [hunk 12]
```diff
@@ -156,6 +151,35 @@ func setupKubeconfig(h *host.Host, c *config.ClusterConfig, n *config.Node, clus
...
+	var err error
+	if driver.IsKIC(h.DriverName) {
+		// for kic drivers we use 127.0.0.1 instead of node IP,
+		// because of Docker on MacOs limitations for reaching to container's IP.
+		hostname = oci.DefaultBindIPV4
+		port, err = oci.HostPortBinding(h.DriverName, h.Name, port)
+		if err != nil {
+			return "", errors.Wrap(err, "host port binding")
+		}
+		if cc.KubernetesConfig.APIServerName != constants.APIServerName {
```
Comment (tstromberg) [inline pkg/minikube/node/config.go]: Can we drop this duplicate check? I just know it's going to get refactored out some day.

```
if driver.IsKIC() {
  hostname =
  port = 
} else {
  hostname = h.Driver.GetIP()
}

if cc.KubernetesConfig.APIServerName != constants.APIServerName {
  hostname = cc.KubernetesConfig.APIServerName
}
return fmt.Sprintf("https://" + net.JoinHostPort(hostname, strconv.Itoa(n.Port))), nil
```

Then you can eliminate the duplicate APIServerName check as well as the duplicate URL generation.

File: pkg/provision/ubuntu.go [hunk 13]
```diff
@@ -155,6 +155,11 @@ WantedBy=multi-user.target
 		return nil, err
 	}
 
+	// because in kic base image we pre-install docker it already has a service file. we need to daemon-reload for the new systemd file
+	if err := p.Service("", serviceaction.DaemonReload); err != nil {
```
Comment (tstromberg) [inline pkg/provision/ubuntu.go]: Can we also do this in buildroot.go?

Otherwise we'll run into a similar problem when we update Docker configs to be different than a previously run.

Comment (medyagh) [reply pkg/provision/ubuntu.go]: ok :)
